crates/rpc/src/cache.rs

```
@@ -0,0 +1,148 @@
+#[derive(Debug, Clone)]
+pub struct TxCache {
+    url: reqwest::Url,
+    client: reqwest::Client,
+}
+
+/// A bundle response from the transaction cache, containing a UUID and a
+/// [`SignetEthBundle`].
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SignetEthBundleResponse {
+    /// The bundle id (a UUID)
+    pub id: String,
+    /// The bundle itself
+    pub bundle: SignetEthBundle,
+}
+
+/// Response from the transaction cache `bundles` endpoint, containing a list of bundles.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct TxCacheBundleResponse {
+    /// the list of bundles
+    pub bundles: Vec<SignetEthBundleResponse>,
+}
+
+/// Response from the transaction cache `transactions` endpoint, containing a list of transactions.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct TxCacheTransactionsResponse {
+    transactions: Vec<TxEnvelope>,
+}
+
+/// Response from the transaction cache `orders` endpoint, containing a list of signed orders.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct TxCacheOrderResponse {
+    orders: Vec<SignedOrder>,
+}
+
+impl TxCache {
+    /// Create a new cache with the given URL and client.
+    pub const fn new_with_client(url: reqwest::Url, client: reqwest::Client) -> Self {
+        Self { url, client }
+    }
+
+    /// Instantiate a new cache with the given URL and a new reqwest client.
+    pub fn new(url: reqwest::Url) -> Self {
+        Self { url, client: reqwest::Client::new() }
+    }
+
+    async fn forward_inner<T: Serialize + Send>(
+        &self,
+        join: &'static str,
+        obj: T,
+    ) -> EthResult<()> {
+        // Append the path to the URL.
+        let url = match self.url.join(join) {
+            Ok(url) => url,
+            Err(e) => {
+                warn!(%e, "Failed to join URL. Not forwarding transaction.");
+                return Ok(());
+            }
+        };
+
+        // Send the object.
+        let _ = self
```

do we want to ignore all errors here? Or do we want to return http errors and let the caller decide what to do?

---

actually, this function currently has no error path at all. so it returns a result, but will never return `Err(_)`

---

ah hmm. do we want to change that?

---

I think that we want to change that. it should return errors instead of ignoring them

---

is this a reasonable way? or do we want to implement custom SignetErrors for this? https://github.com/init4tech/signet-sdk/pull/39/commits/e10f9827b4b1c1df13c41b70498fb9f4554845e3

---

or maybe even have custom TransactionCacheError and implement SignetError from TransactionCacheError..?

---

so assuming the URL is good, the only errors that can be thrown here are `reqwest::Error`. so typically we would eliminate the url angle and then have this function just return reqwest::Error